ethstorage/scanner/worker.go

```diff
@@ -0,0 +1,214 @@
+			if err == es.ErrCommitMismatch {
+				sts.mismatched = append(sts.mismatched, kvIndex)
+				if err := s.fixKv(kvIndex, commit); err != nil {
+					sts.failed = append(sts.failed, kvIndex)
+					s.lg.Error("Scanner: fix blob error", "kvIndex", kvIndex, "err", err)
+					sendError(kvIndex, fmt.Errorf("failed to fix blob: %w", err))
+				} else {
+					sts.fixed = append(sts.fixed, kvIndex)
+					// remove the error from the cache
+					sendError(kvIndex, nil)
+				}
+			} else {
+				sendError(kvIndex, fmt.Errorf("failed to read blob: %w", err))
+			}
+		}
+	}
+	s.nextKvIndex = nextKvIndex
+	if len(kvsInBatch) > 0 {
+		s.lg.Info("Scanner: scan batch done", "from", kvsInBatch[0], "to", kvsInBatch[len(kvsInBatch)-1], "count", len(kvsInBatch), "nextKvIndex", nextKvIndex)
+	}
+	return &sts, nil
+}
+
+func (s *Worker) determineBatchIndexRange(localKvs []uint64) ([]uint64, uint64, error) {
+	localKvTotal := uint64(len(localKvs))
+
+	batchStart := s.nextKvIndex
+	if batchStart >= localKvTotal {
+		batchStart = 0
+		s.lg.Info("Scanner: scan batch start over")
+	}
+	batchEnd := batchStart + uint64(s.cfg.BatchSize)
+	if batchEnd > localKvTotal {
+		batchEnd = localKvTotal
+	}
+	return localKvs[batchStart:batchEnd], batchEnd, nil
+}
+
+func (s *Worker) queryLocalKvs() ([]uint64, error) {
+	kvEntryCount, err := s.l1.GetStorageLastBlobIdx(rpc.LatestBlockNumber.Int64())
+	if err != nil {
+		return nil, fmt.Errorf("failed to query total KV entries: %w", err)
+	}
+	s.lg.Info("Scanner: query KV entry count done", "kvEntryCount", kvEntryCount)
+	var localKvs []uint64
+	kvEntries := s.sm.KvEntries()
+	localShards := s.sm.Shards()
+	s.lg.Info("Scanner: local shards", "shards", localShards)
+	sort.Slice(localShards, func(i, j int) bool {
+		return localShards[i] < localShards[j]
+	})
+	for _, shardId := range localShards {
+		for k := uint64(0); k < kvEntries; k++ {
+			kvIdx := shardId*kvEntries + k
+			if kvIdx < kvEntryCount {
+				// only check non-empty data
+				localKvs = append(localKvs, kvIdx)
+			}
+		}
+	}
+	return localKvs, nil
+}
+
+func (s *Worker) fixKv(kvIndex uint64, commit common.Hash) error {
+	s.lg.Info("Scanner: try to fix blob", "kvIndex", kvIndex, "commit", commit)
+	commitToFetchBlob := commit
+	newCommit, err := s.sm.CheckMeta(kvIndex)
+	if err != nil {
+		if errors.Is(err, es.ErrCommitMismatch) {
+			// the commit in the contract has been updated
+			commitToFetchBlob = newCommit
+		} else {
+			// use the old commit if check meta fails
+			s.lg.Warn("Scanner: check meta error", "kvIndex", kvIndex, "err", err)
+		}
+	} else {
+		// the commit in the contract is the same as the one in the storage
+		if s.cfg.Mode == modeCheckMeta {
+			s.lg.Info("Scanner: KV already fixed by downloader", "kvIndex", kvIndex)
+			return nil
+		}
+		// in case the commit in the contract is same with the one in the storage, and both updated lately
+		commitToFetchBlob = newCommit
+	}
+	blob, err := s.fetchBlob(kvIndex, commitToFetchBlob)
```

Comment:
Can we move this part into TryWriteWithMetaCheck?

Reply:
fixed